[Woo POS] Add remote and logic to load POS products

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #12837
Closes #13068 by switching the providence of data from local storage to the network

Please note that `692` lines are just the JSON simulated network response, so the PR is just ~200 lines 😅 

## Description
This PR updates the `POSItemProvider` to add a new method which fetches 100 products directly from the remote, rather than relying in the app's storage.

Products are saved in-memory for the duration of the POS session as its lifecycle is tied to the view model, only requiring a re-fetch when we exiting and re-entering POS mode.

![Simulator Screen Recording - iPad Pro (12 9-inch) (6th generation) - 2024-06-24 at 14 22 32](https://github.com/woocommerce/woocommerce-ios/assets/3812076/a2551f3b-5b7a-4231-8296-0a4be944a26c)

## Changes

- Add `loadAllSimpleProductsForPointOfSale(_:)` method to `ProductsRemote`. This fetches the first 100 simple products from the store.
- Add `products-load-all-type-simple.json` for testing the simulated response
- Update `PointOfSaleDashboardViewModel` so we inject the item provider directly, rather than the array of items. This also moves the responsibility of the data provenance (networking vs storage) to the view model.

## Question
I've seen a sneaky bug with the "loading" indicator, where it doesn't seem to disappear despite the state changing correctly, the API response being 200, and having all available items in memory, but I'm unable to reproduce at will and only has happened 1/5 times, and only when app is restarted, never on app-pos mode switch: Do you see any reason why the `ItemListView` wouldn't load the correct subview when `isSyncingItems = false`? 

## Testing
- In POS
- Observe a "loading..." text and that only simple products are loaded (bonus points if your store has 25+, you can use https://indiemelon.mystagingwebsite.com/ for this if needed)
- Moving to the checkout flow and going back does not re-fetch the products
- Exiting the POS mode and re-entering re-fetches the products

iamgabrielma committed:
Add remote & tests for fetching POS products

iamgabrielma committed:
Rename providePointOfSaleItemsFromStorage

iamgabrielma committed:
Inject credentials into POSProductProvider

iamgabrielma committed:
Extract and reuse mapper from Product to POSItem

iamgabrielma committed:
Inject itemProvider rather than items in vm

By passing the itemProvider as a dependency, we leave the responsibility of fetching data from remote, storage, or anywhere else to the internals. It also allows us to use async await inside the view model without interfering with the initialization.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Adds ActivityIndicator while syncing products

iamgabrielma committed:
Update test names

iamgabrielma committed:
Lint

iamgabrielma committed:
remove print statement

iamgabrielma committed:
Remove usage of ActivityIndicator

It seems to be a bug with the usage of ActivityIndicator, and -sometimes- the first time we run the app this never disappears despite getting a state switch from the viewmodel and a 200 response from the API. Removing this seems to resolve it, I’m not entirely sure why.

iamgabrielma committed:
Make tests compile

iamgabrielma commented on WooCommerce/Classes/POS/Utils/PreviewHelpers.swift:
```diff
@@ -18,7 +18,11 @@ struct POSProductPreview: POSItem {
+        []
+    }
+
+    func providePointOfSaleItemsFromStorage() -> [POSItem] {
```
We could remove `providePointOfSaleItemsFromStorage`, since we're just fetching from the remote. I've kept it around for faster testing but it's redundant now. Thoughts?

joshheald replied:
If we're not using it, it's clearer to delete it.

iamgabrielma replied:
Sounds good, definitely agree. Removed on https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/3bc796a7429ed84e415bc92cbedb4684aacd6d5e and https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/7f402896355cae566398c262fc43a1513178977e

joshheald commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:
```diff
@@ -73,16 +73,16 @@ final class HubMenuViewModel: ObservableObject {
     private(set) lazy var posItemProvider: POSItemProvider = {
         let storageManager = ServiceLocator.storageManager
         let siteID = ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0
+        let credentials = ServiceLocator.stores.sessionManager.defaultCredentials
```
These are available in the `HubMenuViewModel.credentials` property now, as is the siteID. It would probably be better not to have this as a lazy var as it often leads to issues...

iamgabrielma replied:
> These are available in the HubMenuViewModel.credentials property now, as is the siteID.

Oh nice, updated on [1638abd](https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/1638abdddec131f39ae135779afd78a4c010dc88)

> It would probably be better not to have this as a lazy var as it often leads to issues...

I see 🤔 , for this I believe we've used the same pattern as other view models within the HubMenu. I'll explore this separately and see if/how we can remove the lazy instantiation.

joshheald replied:
Since it gets injected to `PointOfSaleEntryPointView`, and we're unlikely to write tests for that, you could build the default one in the `PointOfSaleEntryPointView.init` function, which would make it a lot simpler.

joshheald commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -9,11 +12,23 @@ public final class POSProductProvider: POSItemProvider {
+        self.init(storageManager: storageManager,
+                  siteID: siteID,
+                  currencySettings: currencySettings,
+                  network: AlamofireNetwork(credentials: credentials))
```
I don't think we're using the `network` any more... if that's going to stay true, we can clean it up along with some imports...

iamgabrielma replied:
I think you meant `storageManager`? Removed on https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/7f402896355cae566398c262fc43a1513178977e

joshheald replied:
I did mean network, but I was wrong about not us using it!

Thanks for cleaning up `storageManager` though.

joshheald commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -25,9 +40,38 @@ public final class POSProductProvider: POSItemProvider {
     /// Provides a`[POSProduct]`array by mapping  simple, purchasable-only Products from storage
     ///
-    public func providePointOfSaleItems() -> [POSItem] {
+    public func providePointOfSaleItemsFromStorage() -> [POSItem] {
```
See previous comments about getting rid/renaming stuff 😊

iamgabrielma replied:
Yep! Done on https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/3bc796a7429ed84e415bc92cbedb4684aacd6d5e

joshheald commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -25,9 +40,38 @@ public final class POSProductProvider: POSItemProvider {
+    private func mapProductsToPOSItems(products: [Product]) -> [POSItem] {
+        let currencyFormatter = CurrencyFormatter(currencySettings: currencySettings)
+        return products.map { product in
+            let formattedPrice = currencyFormatter.formatAmount(product.price, with: currencySettings.currencyCode.rawValue) ?? "-"
```
Does the code need to be provided, since we already created the formatter with this same set of currency settings?

iamgabrielma replied:
I'm not sure to understand this one, can you clarify?

joshheald replied:
I think you can just do
```suggestion
            let formattedPrice = currencyFormatter.formatAmount(product.price) ?? "-"
```
because you already gave the `currencyFormatter` these `currencySettings`. Take a look at the `formatAmount` implementation and optional params.

iamgabrielma replied:
I see, thanks! Addressed here: https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/4a40286979fd8a63ad3db548da99e21efb890b3f

joshheald commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -25,9 +40,38 @@ public final class POSProductProvider: POSItemProvider {
+            let thumbnailSource = product.images.first?.src
+            let productCategories = product.categories.map { $0.name }
+
+            return POSProduct(itemID: UUID(),
```
Is this UUID going to cause us issues, since it won't be stable? Can we just use productID here, or a combination of product and siteID? It might not matter.

iamgabrielma replied:
In this case I think is not enough with `productID`, since for example the `CartView` might render the same product multiple times, which causes issues to SwiftUI when cannot identify the elements uniquely only by `productID` and needs other unique identifier, eg:
```
ForEach(viewModel.itemsInCart, id: \.id)
```

I think creating a new UUID should be fine (for now), but I'll check if we may want to use some sort combination for a more stable unique ID.

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -53,23 +54,39 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
+    private func populatePointOfSaleItems() {
+        Task { @MainActor in
+            do {
+                items = try await itemProvider.providePointOfSaleItemsFromNetwork()
```
When we remove the `FromStorage` variant, let's rename this back to `fetchItems()`. There's no reason for the view model to know how its item provider works...

You could even make it a var, and have the items be fetched when it's created, adding a refresh function as well. Better to avoid having a var do the fetch though, not least because it's async.

iamgabrielma replied:
> You could even make it a var, and have the items be fetched when it's created, adding a refresh function as well. Better to avoid having a var do the fetch though, not least because it's async.

Thanks for the tip! I'll check this when adding the pull-to-refresh here https://github.com/woocommerce/woocommerce-ios/issues/12837

joshheald reviewed (approved):
Some changes – main one is the design of the provider.

It's best with these kinds of classes to keep the words in the function names generic unless you absolutely have to be specific. The viewmodel shouldn't have to care whether the items come from the network or storage.

In this case, I don't see a reason for this class to even keep the storage query capability... but if we did, that should be an implementation detail private to the provider, not something the view model has to decide – we should be able to provide an itemprovider that works totally differently but has the same interface and still have it work.

> I've seen a sneaky bug with the "loading" indicator, where it doesn't seem to disappear despite the state changing correctly, the API response being 200, and having all available items in memory, but I'm unable to reproduce at will and only has happened 1/5 times, and only when app is restarted, never on app-pos mode switch: Do you see any reason why the ItemListView wouldn't load the correct subview when isSyncingItems = false?

I'm not 100% sure of the reason, but I'd look at how and when the `populatePointOfSaleItems` function is called. Doing it from init means it's likely to happen before the view (or anything else) is subscribed to the `isSyncingItems` publisher. Yes, it would make sense for SwiftUI to get the latest value of the publisher when it subscribes the view, and maybe that's what happens... but also maybe it's not. SwiftUI can be tricksy about that kind of stuff.

Note that I've not checked whether that's actually the reason – the view might get the expected `isSyncingItems = false` when it's created and just not respond to it for some reason. That just seems less likely.

I _did_ try making the `populatePointOfSaleItems` callable from a `.task` modifier by changing its implementation to 

```
    @MainActor
    func populatePointOfSaleItems() async {
        do {
            items = try await itemProvider.providePointOfSaleItemsFromNetwork()
            isSyncingItems = false
        } catch {
            debugPrint("\(error)")
        }
    }
```

Doing that, and calling it from a `task` modifier on the root `PointOfSaleDashboardView` instead of the init function made it work exactly as expected. That doesn't mean it's the best option, just something I quickly tried!

There are probably also ways to separate the change to `isSyncingItems` from the network response itself, so that we don't change it until the views are hooked up but still do the network request earlier... but in this case I think that would be unnecessary complexity.

iamgabrielma committed:
Remove fetching from storage. Collapse into providePointOfSaleItems func

iamgabrielma committed:
Remove unnecessary property declarations

iamgabrielma committed:
Move call to populate products to task in View

This seems to resolve the loading indicator bug not disappearing in some instances due to -possibly- the view not getting the latest value of the published state after fetching items

iamgabrielma committed:
Remove StorageManagerType dependency

iamgabrielma commented:
Thanks for the review!

> I did try making the populatePointOfSaleItems callable from a .task modifier

I'm not sure either, but trying the approach you shared on https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/97a969a0f6d78e5946d3ece32b750ed68e1cac41 solves the problem indeed, and its populated exactly as expected, I cannot see the problem anymore 🤔 I'll time-box some additional time to investigate and add further tests to see if something comes up.

joshheald commented on WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift:
```diff
@@ -12,7 +12,7 @@ final class PointOfSaleDashboardViewModelTests: XCTestCase {
         super.setUp()
         cardPresentPaymentService = CardPresentPaymentPreviewService()
-        sut = PointOfSaleDashboardViewModel(items: [],
+        sut = PointOfSaleDashboardViewModel(itemProvider: POSItemProviderPreview(),
```
We should make a mock object here – even though this is convenient, things should have one job, and with this we're expecting the Preview provider to be as simple as possible _and_ a test double which we can check when we add more tests for the view model..

iamgabrielma replied:
Right! I've created a `MockPOSItemProvider` on a previous commit but I realized this was sitting in `Yosemite` and testing the mock, rather than the implementation. I've moved it to `WooCommerce` here: https://github.com/woocommerce/woocommerce-ios/pull/13116/commits/3b51e7ab938865fd7537d7e901bcdc7f3d3fd466 as an extension for now, will extract it as soon as we need to add further tests.

iamgabrielma committed:
Remove unnecessary formatter details

The currency code is already provided as default value in the `formatAmount()` method, no need to pass it through.

iamgabrielma committed:
Address usage of MockPOSItemProvider in tests
